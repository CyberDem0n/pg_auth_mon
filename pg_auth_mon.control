comment = 'report failed authentication attempt to a separate log file'
default_version = '1.0'
module_pathname = '$libdir/pg_auth_mon'
relocatable = true
