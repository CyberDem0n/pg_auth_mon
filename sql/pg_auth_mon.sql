create role auth_super with superuser login password 'postgres';
create role auth_nologin with password 'postgres';
create role auth_test with login password 'wrongpassword';

create database testdb;

create extension pg_auth_mon;

--1.Successful Login attempt
\! PGPASSWORD=postgres psql -U auth_super -d testdb -c "select 1"
select rolname, successful_attempts, total_hba_conflicts, other_auth_failures from pg_auth_mon where rolname like 'auth_%';

--2.Login attempt by invalid username
\! PGPASSWORD=postgres psql -U auth_blah -d testdb  -c "select 1"
select rolname, uid, successful_attempts, total_hba_conflicts, other_auth_failures from pg_auth_mon where uid = 0;

--3.Login attempt by username who is not allowed to login is not authentication failure
\! PGPASSWORD=postgres psql -U auth_nologin -d testdb -c "select 1"
select rolname, successful_attempts, total_hba_conflicts, other_auth_failures from pg_auth_mon where rolname like 'auth_%';

--4.Login attempt by a valid user with a wrong password 
\! PGPASSWORD=postgres psql -U auth_test -d testdb -c "select 1"
select rolname, successful_attempts, total_hba_conflicts, other_auth_failures from pg_auth_mon where rolname like 'auth_%' order by rolname;

--Cleanup
drop role auth_nologin;
drop role auth_test;
drop database testdb;
drop role auth_super;
