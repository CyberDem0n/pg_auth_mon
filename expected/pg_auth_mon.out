create role auth_super with superuser login password 'postgres';
create role auth_nologin with password 'postgres';
create role auth_test with login password 'wrongpassword';
create database testdb;
create extension pg_auth_mon;
--1.Successful Login attempt
\! PGPASSWORD=postgres psql -U auth_super -d testdb -c "select 1"
 ?column? 
----------
        1
(1 row)

select rolname, successful_attempts, total_hba_conflicts, other_auth_failures from pg_auth_mon where rolname like 'auth_%';
  rolname   | successful_attempts | total_hba_conflicts | other_auth_failures 
------------+---------------------+---------------------+---------------------
 auth_super |                   1 |                   0 |                   0
(1 row)

--2.Login attempt by invalid username
\! PGPASSWORD=postgres psql -U auth_blah -d testdb  -c "select 1"
psql: error: could not connect to server: FATAL:  password authentication failed for user "auth_blah"
select rolname, uid, successful_attempts, total_hba_conflicts, other_auth_failures from pg_auth_mon where uid = 0;
 rolname | uid | successful_attempts | total_hba_conflicts | other_auth_failures 
---------+-----+---------------------+---------------------+---------------------
         |   0 |                   0 |                   0 |                   1
(1 row)

--3.Login attempt by username who is not allowed to login is not authentication failure
\! PGPASSWORD=postgres psql -U auth_nologin -d testdb -c "select 1"
psql: error: could not connect to server: FATAL:  role "auth_nologin" is not permitted to log in
select rolname, successful_attempts, total_hba_conflicts, other_auth_failures from pg_auth_mon where rolname like 'auth_%';
   rolname    | successful_attempts | total_hba_conflicts | other_auth_failures 
--------------+---------------------+---------------------+---------------------
 auth_nologin |                   1 |                   0 |                   0
 auth_super   |                   1 |                   0 |                   0
(2 rows)

--4.Login attempt by a valid user with a wrong password 
\! PGPASSWORD=postgres psql -U auth_test -d testdb -c "select 1"
psql: error: could not connect to server: FATAL:  password authentication failed for user "auth_test"
select rolname, successful_attempts, total_hba_conflicts, other_auth_failures from pg_auth_mon where rolname like 'auth_%' order by rolname;
   rolname    | successful_attempts | total_hba_conflicts | other_auth_failures 
--------------+---------------------+---------------------+---------------------
 auth_nologin |                   1 |                   0 |                   0
 auth_super   |                   1 |                   0 |                   0
 auth_test    |                   0 |                   0 |                   1
(3 rows)

--Cleanup
drop role auth_nologin;
drop role auth_test;
drop database testdb;
drop role auth_super;
